### PR TITLE
Add Image-Recognition and Object-Detection TensorFlow processors

### LINKED ISSUE #2: Add Image-Recognition and Object-Detection TensorFlow processors
1. TensorFlow `image-recognition` processor  based on the state-of-the-art [Inception model](https://arxiv.org/abs/1512.00567). It extends and customizes the core `tensorflow-processor`. 
The [Inception Github project](https://github.com/tensorflow/models/tree/master/inception) project implements an [image recognition model](https://www.tensorflow.org/tutorials/image_recognition) model. Later allows existing, pre-trained inception models to be [customized](https://www.tensorflow.org/tutorials/image_retraining) for specific domain or one can train it with custom image datasets.

2. Add [Tensorflow Object Detection API](https://github.com/tensorflow/models/tree/master/research/object_detection) processor

### PR DESCRIPTION
**Pull request by tzolov**
Resolves #2

 - Align with Spring Boot 2.0, Spring Cloud 2.0 and SCSt 2.0 stack
 - Upgrade to Tensorflow 1.6.0-rc1
 - Add Image-Recognition processor
 - Add Object Detection API processor + Protobuf pre-generated labels reader
 - Extend the common Tensorflow Processor to support multiple evaluation outputs. (FetchModel parameter allow comma-separated out model names)
   - Fix the processor output name (in header or tuple) to "result". Use to match the fetch name
   - Remove the redundant fetch-index property
   - Replace @ServiceActivator by @ServiceListener
 - Refactor the contentType handling to align it with new SCSt 2.0 requirements

### CODE REVIEW

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/DefaultOutputMessageBuilder.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	private String findContentType(Object computedScore) {
+		if (computedScore instanceof Tuple) {
+			return APPLICATION_X_SPRING_TUPLE;
+		}
+		else if (computedScore instanceof String) {
+			return MimeTypeUtils.APPLICATION_JSON_VALUE;
+		}
+
+		// default
+		return MimeTypeUtils.APPLICATION_JSON_VALUE;
```
Comment (artembilan) [inline spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/DefaultOutputMessageBuilder.java]: Hm. This default is the same as a variant for the `String`.  
Maybe something is missed in the logic?

File: spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+		try {
+			if (result != null) {
+				BufferedImage originalImage = ImageIO.read(new ByteArrayInputStream(imageBytes));
+
+				Graphics2D g = originalImage.createGraphics();
+				g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
+				FontMetrics fm = g.getFontMetrics();
+
+				Tuple resultTuple = new JsonStringToTupleConverter().convert(result.toString());
+				ArrayList<Tuple> labels = (ArrayList) resultTuple.getValues().get(0);
```
Comment (artembilan) [inline spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java]: image


Would be better to have generic args and use `@SuppressWarning` on the method.

File: spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+				baos.flush();
+				imageBytes = baos.toByteArray();
+				baos.close();
+			}
+		}
+		catch (IOException e) {
+			logger.error(e);
+		}
+
+		// Null mend that QR image is found and not output message will be send.
```
Comment (artembilan) [inline spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java]: Null **means** ?
and **no** output  ?

File: spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/GraphicsUtils.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+package org.springframework.cloud.stream.app.object.detection.processor;
+
+
+import java.awt.*;
+import java.awt.geom.Rectangle2D;
+import java.awt.image.BufferedImage;
+
+/**
+ * @author Christian Tzolov
```
Comment (artembilan) [inline spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/GraphicsUtils.java]: May we have some JavaDocs on this class to determine the reason of such a big class? 
Thanks

File: spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/GraphicsUtils.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.app.object.detection.processor;
+
+
+import java.awt.*;
```
Comment (artembilan) [inline spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/GraphicsUtils.java]: No asterisk imports, please.

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/InvalidTupleTensorflowEncoding.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.cloud.stream.app.tensorflow.processor;
+
+/**
+ * @author Christian Tzolov
+ */
+public class InvalidTupleTensorflowEncoding extends RuntimeException {
```
Comment (artembilan) [inline spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/InvalidTupleTensorflowEncoding.java]: Yeah... Exception must have an `Exception` suffix to be clear...
But that's IMHO

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java [hunk 7]
```diff
@@ -56,18 +62,14 @@
...
- * By default the inference result is returned in the outbound Message payload. If the saveResultInHeader property is
- * set to true then the inference result would be stored in the outbound Message header by name as set by
- * the getResultHeader property. In this case the message payload is the same like the inbound message payload.
+ * The {@link OutputMessageBuilder} defines how the computed inference score is arranged withing the output message.
  *
  * @author Christian Tzolov
  * @author Artem Bilan
  */
 @EnableConfigurationProperties(TensorflowCommonProcessorProperties.class)
 public class TensorflowCommonProcessorConfiguration implements AutoCloseable {
```
Comment (artembilan) [inline spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java]: We don't need to have here `AutoCloseable`. 
Since you declare `TensorFlowService` as a bean, Spring will take care about its own `AutoCloseable` already.

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java [hunk 8]
```diff
@@ -146,29 +135,51 @@ public TensorFlowService tensorFlowService() throws IOException {
...
+			}
+
+			throw new MessageConversionException("Unsupported input format: " + input);
+		};
+	}
 
+	@Bean
+	@StreamMessageConverter
+	public MessageConverter toTensorMessageConverter() {
+		return new AbstractMessageConverter(new MimeType("application", "x-tensor")) {
```
Comment (artembilan) [inline spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java]: Where this converter is used?

I thought we need them when we convert `input`/`output` from/to external world, therefore it is a bit confusing for me why would one need this custom mime-type?

Thanks

File: spring-cloud-starter-stream-processor-image-recognition/src/test/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionProcessorPropertiesTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.test.util.EnvironmentTestUtils;
+import org.springframework.boot.test.util.TestPropertyValues;
+import org.springframework.context.annotation.AnnotationConfigApplicationContext;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * @author Christian Tzolov
+ */
+public class ImageRecognitionProcessorPropertiesTest {
```
Comment (artembilan) [inline spring-cloud-starter-stream-processor-image-recognition/src/test/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionProcessorPropertiesTest.java]: Must be plural - `Tests`

File: spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/ObjectDetectionTensorflowOutputConverter.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+			// - 1 as the first dimension
+			// - maxObjects as the second dimension
+			// While boxesT will have 4 as the third dimension (2 sets of (x, y) coordinates).
+			// This can be verified by looking at scoresT.shape() etc.
+			int maxObjects = (int) scoresTensor.shape()[1];
+			float[] scores = scoresTensor.copyTo(new float[1][maxObjects])[0];
+			float[] classes = classesTensor.copyTo(new float[1][maxObjects])[0];
+			float[][] boxes = boxesTensor.copyTo(new float[1][maxObjects][4])[0];
+
+			// Work in progress
```
Comment (artembilan) [inline spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/ObjectDetectionTensorflowOutputConverter.java]: Should we expect something else here to be done?

Thanks

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/DefaultOutputMessageBuilder.java (hunk 1 above)
Comment (tzolov) [reply spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/DefaultOutputMessageBuilder.java]: +1

File: spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java (hunk 2 above)
Comment (tzolov) [reply spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java]: +1. I've replaced with `ArrayList<Tuple> labels  = resultTuple.getValue("labels", ArrayList.class);`

File: spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java (hunk 3 above)
Comment (tzolov) [reply spring-cloud-starter-stream-processor-image-recognition/src/main/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionOutputMessageBuilder.java]: +1

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/InvalidTupleTensorflowEncoding.java (hunk 6 above)
Comment (tzolov) [reply spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/InvalidTupleTensorflowEncoding.java]: +1

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java (hunk 7 above)
Comment (tzolov) [reply spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java]: +1

File: spring-cloud-starter-stream-processor-image-recognition/src/test/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionProcessorPropertiesTest.java (hunk 9 above)
Comment (tzolov) [reply spring-cloud-starter-stream-processor-image-recognition/src/test/java/org/springframework/cloud/stream/app/image/recognition/processor/ImageRecognitionProcessorPropertiesTest.java]: +1

File: spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/ObjectDetectionTensorflowOutputConverter.java (hunk 10 above)
Comment (tzolov) [reply spring-cloud-starter-stream-processor-object-detection/src/main/java/org/springframework/cloud/stream/app/object/detection/processor/ObjectDetectionTensorflowOutputConverter.java]: This is to enable support for image segmentation feature: https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/instance_segmentation.md
While the TF models already support this and we can read the metadata, it turned to be quite involving to convert the mask metadata into mask images.
So this fragment can be regarded as an enabler to support the image segmentation in some of the following up releases. 
I will improve the  descriptions.

File: spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java (hunk 8 above)
Comment (tzolov) [reply spring-cloud-starter-stream-common-tensorflow/src/main/java/org/springframework/cloud/stream/app/tensorflow/processor/TensorflowCommonProcessorConfiguration.java]: Good spot. I believe this is a left over from my experiments with the new content-type support. I will remove it

**Comment by tzolov** [comment]
Thanks @artembilan! I've pushed an update to address the review comments. 
Also note i've not disabled the integration tests for the Twitter Sentiment, Image Recognition and Object Detection processors. Later pulls large (> 50 MB), pre-build TF models from bintray.

File: spring-cloud-starter-stream-processor-tensorflow/README.adoc [hunk 11]
```diff
@@ -1,11 +1,32 @@
...
+minimal_graph = convert_variables_to_constants(sess, sess.graph_def, ['<model output>'])
+tf.train.write_graph(minimal_graph, SAVE_DIR, 'my_graph.proto', as_text=False)
+tf.train.write_graph(minimal_graph, SAVE_DIR, 'my.txt', as_text=True)
+```
+
 image::{image-root}/TensorFlowProcessorArcutectureOverview.png[]
 
+The `--tensorflow.model` property configures the Processor with the location of the serialized Tensorflow model.
+
+The `TensorflowInputConverter` implementations converts the input data into the format, specific for the given model.
```
Comment (artembilan) [inline spring-cloud-starter-stream-processor-tensorflow/README.adoc]: I think to align with the next sentence the `implementations` word is redundant.

File: spring-cloud-starter-stream-common-tensorflow/pom.xml [hunk 12]
```diff
@@ -21,6 +21,14 @@
             <groupId>org.tensorflow</groupId>
             <artifactId>proto</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.tensorflow</groupId>
+            <artifactId>libtensorflow</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.tensorflow</groupId>
+            <artifactId>libtensorflow_jni_gpu</artifactId>
```
Comment (artembilan) [inline spring-cloud-starter-stream-common-tensorflow/pom.xml]: Would you mind to elaborate more here, please?

Looks like using this here we don't have choice in the target application: https://www.tensorflow.org/versions/master/install/install_linux#determine_which_tensorflow_to_install

I wonder what is going to happen if I don't have the proper GPU or I am just on Windows (although my Nvidia GPU is pretty good :smile: )

I mean shouldn't we leave such a choice for end-user?
Or doesn't it just have any effect if we don't meet GPU requirements and it just fallback to CPU mode?

Thanks

File: spring-cloud-starter-stream-processor-tensorflow/README.adoc (hunk 11 above)
Comment (tzolov) [reply spring-cloud-starter-stream-processor-tensorflow/README.adoc]: +1

File: spring-cloud-starter-stream-common-tensorflow/pom.xml (hunk 12 above)
Comment (tzolov) [reply spring-cloud-starter-stream-common-tensorflow/pom.xml]: Good point!
According to this doc: https://www.tensorflow.org/install/install_java 
> GPU acceleration is available via Maven only for Linux and only if your system meets the requirements for GPU.
So my assumption is that those dependencies will only be activated if the above conditions hold. For example it works find on my MacBook laptop using only the CPU there (e.g. the GPU dependencies are ignored).
Also with those additional maven dependencies i hoped we could address issues like this: https://stackoverflow.com/questions/46430512/tensorflow-train-a-model-with-gpus-and-inference-with-java-api

But having said this i acknowledge that i have not tested this functionality, so i will comment out those dependencies and open a separate issue for tackling this.

**Comment by artembilan** [comment]
Merged as https://github.com/spring-cloud-stream-app-starters/tensorflow/commit/cd1ea722ea955cabaaf301ff975ff9e2a3fe9e0a
